abci/check_tx.go

```diff
@@ -0,0 +1,90 @@
+package abci
+
+import (
+	"context"
+	"fmt"
+
+	abci "github.com/cometbft/cometbft/abci/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/skip-mev/pob/mempool"
+)
+
+type (
+	// CheckTx is baseapp's CheckTx method that checks the validity of a
+	// transaction.
+	CheckTx func(abci.RequestCheckTx) abci.ResponseCheckTx
+
+	// GetContextForBidTx is an interface that allows us to get a context for
+	// a bid transaction.
+	GetContextForBidTx func(req abci.RequestCheckTx) sdk.Context
+
+	// CheckTxMempool is an interface that allows us to check if a transaction
+	// exists in the mempool and get the bid info of a transaction.
+	CheckTxMempool interface {
+		GetAuctionBidInfo(tx sdk.Tx) (*mempool.AuctionBidInfo, error)
+		Insert(ctx context.Context, tx sdk.Tx) error
+		WrapBundleTransaction(tx []byte) (sdk.Tx, error)
+	}
+
+	// BaseApp is an interface that allows us to call baseapp's CheckTx method.
+	BaseApp interface {
+		CheckTx(abci.RequestCheckTx) abci.ResponseCheckTx
+	}
+)
+
+// CheckTxHandler is a wrapper around baseapp's CheckTx method that allows us to
+// verify bid transactions against the latest committed state. All other transactions
+// are executed normally. We must verify each bid tx and all of its bundled transactions
+// before we can insert it into the mempool against the latest commit state because
+// otherwise the auction can be griefed. No state changes are applied to the state
+// during this process.
+func CheckTxHandler(baseApp BaseApp, getContextForBidTx GetContextForBidTx, txDecoder sdk.TxDecoder, mempool CheckTxMempool, anteHandler sdk.AnteHandler) CheckTx {
+	return func(req abci.RequestCheckTx) abci.ResponseCheckTx {
+		sdkTx, err := txDecoder(req.Tx)
```

Comment:
Probably want to set up something with recovery in the case when anything fails (similar to how `CheckTx` and `runTx` does it in base app). Not fully sure if we need to set up the same recovery middleware `runTx` does.